tests/test_execution.py

```diff
@@ -314,3 +314,75 @@ def test_aggregate_composite_traintuples(global_execution_env):
             traintuple=traintuple,
         )
         sessions[0].add_testtuple(spec).future().wait()
+
+
+@pytest.mark.parametrize('fail_count', [5, 6])
+def test_execution_retry_on_fail(fail_count, global_execution_env):
+    """Execution of a traintuple which fails on the N first tries, and suceeds on the N+1th try"""
+
+    # This test ensures the compute task retry mechanism works correctly.
+    #
+    # It executes an algorithm that `raise`s on the first N runs, and then
+    # succeeds.
+    #
+    # /!\ This test should ideally be part of the substra-backend project,
+    #     not substra-tests. For the sake of expendiency, we're keeping it
+    #     as part of substra-tests for now, but we intend to re-implement
+    #     it in substra-backend instead eventually.
+    # /!\ This test makes use of the "local" folder to keep track of a counter.
+    #     This is a hack to make the algo raise or succeed depending on the retry
+    #     count. Ideally, we would use a more elegant solution.
+    # /!\ This test doesn't validate that an error in the docker build phase (of
+    #     the compute task execution) triggers a retry. Ideally, it's that case that
+    #     would be tested, since errors in the docker build are the main use-case
+    #     the retry feature was build for.
+
+    RETRY_ALGO_SNIPPET_TOREPLACE = f"""
+    tools.algo.execute(TestAlgo())"""
+
+    RETRY_ALGO_SNIPPET_REPLACEMENT = f"""
+    counter_path = "/sandbox/local/counter"
+    counter = 0
+    try:
+        with open(counter_path) as f:
+            counter = int(f.read())
+    except IOError:
+        pass # file doesn't exist yet
+
+    # Fail if the counter is below the retry count
+    if counter < <FAIL_COUNT>:
+        counter = counter + 1
+        with open(counter_path, 'w') as f:
+            f.write(str(counter))
+        raise Exception("Intentionally keep on failing until we have failed <FAIL_COUNT> time(s). The algo has now \
+            failed " + str(counter) + " time(s).")
+
+    # The counter is greater than the retry count
+    tools.algo.execute(TestAlgo())"""
```

Comment:
Instead of using a `replace()` you could replace in `RETRY_ALGO_SNIPPET_TOREPLACE` variable, `<FAIL_COUNT>` with `${fail_count}`.

It would be more pythonic and will make use of the f-string.

Reply:
Fixed: https://github.com/SubstraFoundation/substra-tests/pull/39/commits/9b245819deea49adc4ff0be51bf9bed4ce3dc38e